### PR TITLE
PC-597: Creating a master CSV for Consortium leads in addition to individual CSVs

### PR DESCRIPTION
**Pull request by samyou-softwire**
closes PC-597

image

queries which consortiums the user manages and if any exist, it will collate a csv download on the downloads page per month per consortium

for this ticket I defined a consortium lead such that a user is a consortium lead IIF they are a lead for every LA in a consortium, which I believe is how they are implemented on live

some new consortium data was added to the admin portal, yielded from the provided dsa tracker spreadsheet. this includes consortium name, and consortium member LAs. for easy querying a dictionary of consortium to LAs as well as LA to parent consortium was added

to support items on the menu of two different download types an abstract parent class was added for csv files. this contains generic information about the file download and is concretely implemented by consortiums & LAs

the csv files are generated using the added CsvHelper package. the consortium csv files have a new column as well for their yielded consortium as this is not included in the LA csv files

on the frontend, consortiums can be identified as their name ends (Consortium). they appear in the filter, though the ticket did not specify how exactly filtering by consortium should work. for now, the implemented option is that when filtering out the consortium it will only remove the consortium csv, and not do any additional filtering on the LA csv downloads in that consortium

the last download time for a consortium is defined as the earliest download time of its LAs. the last updated time is defined as the latest upload time of its LAs. this seems the most intuitive order

on the backend the last updated and last downloaded stats are calculated from the LA csv files the consortium csv file is based on. there are no database migrations, downloading a combined csv is audited as downloading each LA csv individually

new tests are included

### CODE REVIEW

**Comment by samyou-softwire** [comment]
for generating the consortium data the following kotlin code was written (requiring an input csv of LA information)
[desnz-consortium-parser.zip](https://github.com/UKGovernmentBEIS/desnz-home-energy-retrofit-portal-beta/files/14128502/desnz-consortium-parser.zip)

**Review by jamiehumphries — CHANGES_REQUESTED** [review]
There's quite a lot that it would be good to revisit and tidy up, but I've just put a few comments on bits that we can fix up quickly before this release.

File: HerPortal.BusinessLogic/Models/ConsortiumData.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+﻿namespace HerPortal.BusinessLogic.Models;
+
+public static class ConsortiumData
+{
+    // The mapping from custodian code to name comes from the publicly available "Local custodian codes" download link
+    // on https://www.ordnancesurvey.co.uk/business-government/tools-support/addressbase-support
```
[the remark this answers is not in the document]
Comment (samyou-softwire) [reply HerPortal.BusinessLogic/Models/ConsortiumData.cs:6]: oops

File: HerPortal.BusinessLogic/Models/LocalAuthorityData.cs [hunk 2]
```diff
@@ -375,4 +375,266 @@ public static class LocalAuthorityData
         { "1845", "Wyre Forest" },
         { "2741", "York" },
     };
+    
+    public static readonly Dictionary<string, string> LocalAuthorityConsortiumCodeByCustodianCode = new()
```
[the remark this answers is not in the document]
Comment (samyou-softwire) [reply HerPortal.BusinessLogic/Models/LocalAuthorityData.cs:379]: done

out of interest, any particular reason to now define a separate function to generate the dictionary?

File: HerPortal.BusinessLogic/Services/CsvFileService/AbstractCsvFileData.cs [hunk 3]
```diff
@@ -1,8 +1,8 @@
 namespace HerPortal.BusinessLogic.Services.CsvFileService;
 
-public class CsvFileData
+public abstract class AbstractCsvFileData
```
[the remark this answers is not in the document]
Comment (samyou-softwire) [reply HerPortal.BusinessLogic/Services/CsvFileService/AbstractCsvFileData.cs:3]: is now CsvFileData

**Review by LadunOmideyiSoftwire — APPROVED** [review]

**Pull request merged**